api/v1alpha1/workspace_validation.go

```diff
@@ -35,20 +35,152 @@ func (w *Workspace) Validate(ctx context.Context) (errs *apis.FieldError) {
 			w.Inference.validateUpdate(&old.Inference).ViaField("inference"),
+			w.Tuning.validateUpdate(&old.Tuning).ViaField("tuning"),
+			w.Tuning.Input.validateUpdate(old.Tuning.Input).ViaField("input"),
+			w.Tuning.Output.validateUpdate(old.Tuning.Output).ViaField("output"),
+			w.Resource.validateUpdate(&old.Resource).ViaField("resource"),
 		)
 	}
 	return errs
 }
 
+func (w *Workspace) validateCreate() (errs *apis.FieldError) {
+	inferenceSpecified := w.Inference.Preset != nil || w.Inference.Template != nil
+	tuningSpecified := w.Tuning.Input != nil
+	if inferenceSpecified != tuningSpecified {
+		errs = errs.Also(apis.ErrGeneric("Either Inference or Tuning must be specified, but not both", ""))
+	}
+	return errs
+}
+
+func (w *Workspace) validateUpdate(old *Workspace) (errs *apis.FieldError) {
+	// Check inference specified
+	oldInferenceSpecified := old.Inference.Preset != nil || old.Inference.Template != nil
+	inferenceSpecified := w.Inference.Preset != nil || w.Inference.Template != nil
+	// Check tuning specified
+	oldTuningSpecified := old.Tuning.Input != nil
+	tuningSpecified := w.Tuning.Input != nil
+	if (!oldInferenceSpecified && inferenceSpecified) || (oldInferenceSpecified && !inferenceSpecified) {
+		errs = errs.Also(apis.ErrGeneric("Inference field cannot be toggled once set", "inference"))
+	}
+
+	if (!oldTuningSpecified && tuningSpecified) || (oldTuningSpecified && !tuningSpecified) {
+		errs = errs.Also(apis.ErrGeneric("Tuning field cannot be toggled once set", "tuning"))
+	}
+	return errs
+}
+
+func (r *TuningSpec) validateCreate() (errs *apis.FieldError) {
+	if r.Input == nil {
+		errs = errs.Also(apis.ErrMissingField("Input"))
+	}
+	if r.Output == nil {
+		errs = errs.Also(apis.ErrMissingField("Output"))
+	}
+	// Currently require a preset to specified, in future we can consider defining a template
+	if r.Preset == nil {
+		errs = errs.Also(apis.ErrMissingField("Preset"))
+	}
+	// TODO: We have to register training plugins and check if it preset exists in plugins here
+	methodLowerCase := strings.ToLower(string(r.Method))
+	if methodLowerCase != string(TuningMethodLora) && methodLowerCase != string(TuningMethodQLora) {
+		errs = errs.Also(apis.ErrInvalidValue(r.Method, "Method"))
+	}
+	return errs
+}
+
+func (r *TuningSpec) validateUpdate(old *TuningSpec) (errs *apis.FieldError) {
+	if !reflect.DeepEqual(old.Input, r.Input) {
+		errs = errs.Also(apis.ErrGeneric("Input field cannot be changed", "Input"))
+	}
+	if !reflect.DeepEqual(old.Output, r.Output) {
+		errs = errs.Also(apis.ErrGeneric("Output field cannot be changed", "Output"))
+	}
+	if !reflect.DeepEqual(old.Preset, r.Preset) {
+		errs = errs.Also(apis.ErrGeneric("Preset cannot be changed", "Preset"))
+	}
+	// We will have to consider supporting tuning method and config fields changing
+	methodLowerCase := strings.ToLower(string(r.Method))
+	if methodLowerCase != string(TuningMethodLora) && methodLowerCase != string(TuningMethodQLora) {
```

Comment:
We should not even allow changing the method.

Reply:
done